File: rplugin/python3/denite/source/junkfile.py
# ============================================================================
# FILE: junkfile.py
# AUTHOR: Shougo Matsushita <Shougo.Matsu at gmail.com>
# License: MIT license
# ============================================================================

from .base import Base
from time import strftime
import os


class Source(Base):

    def __init__(self, vim):
        super().__init__(vim)

        self.name = 'junkfile'
        self.kind = 'file'

    def gather_candidates(self, context):
        self.vim.call('junkfile#init')
        base = self.vim.vars['junkfile#directory']

        candidates = []
        if context['args'] and context['args'][0] == 'new':
            context['is_interactive'] = True
            filename = strftime('%Y/%m/%Y-%m-%d-%H%M%S.') + context['input']
            candidates.append({
                'word': os.path.basename(filename),
                'abbr': '[new] ' + os.path.basename(filename),
                'action__path': os.path.join(base, filename),
            })
        else:
            for root, dirs, files in os.walk(base):
                for f in files:
                    candidates.append({
                        'word': f,
                        'action__path': os.path.join(root, f),
                    })
            candidates = sorted(candidates, key=lambda x:
                                os.path.getmtime(x['action__path']),
                                reverse=True)
        return candidates

File: denops/@ddu-sources/junkfile.ts
import { type Item } from "jsr:@shougo/ddu-vim@~6.1.0/types";
import { BaseSource } from "jsr:@shougo/ddu-vim@~6.1.0/source";
import { type ActionData } from "jsr:@shougo/ddu-kind-file@~0.9.0";

import type { Denops } from "jsr:@denops/std@~7.1.0";
import * as vars from "jsr:@denops/std@~7.1.1/variable";
import * as fn from "jsr:@denops/std@~7.1.1/function";

import { join } from "jsr:@std/path@~1.0.3/join";
import { basename } from "jsr:@std/path@~1.0.2/basename";
import { relative } from "jsr:@std/path@~1.0.2/relative";

type Params = Record<string, never>;

export class Source extends BaseSource<Params> {
  override kind = "file";

  override gather(args: {
    denops: Denops;
    sourceParams: Params;
    input: string;
  }): ReadableStream<Item<ActionData>[]> {
    return new ReadableStream({
      async start(controller) {
        await args.denops.call("junkfile#init");

        const dir = await fn.expand(
          args.denops,
          await vars.g.get(args.denops, "junkfile#directory"),
        ) as string;

        const tree = async (root: string) => {
          let items: Item<ActionData>[] = [];

          for await (const entry of Deno.readDir(root)) {
            const path = join(root, entry.name);
            if (!entry.isDirectory) {
              items.push({
                word: relative(dir, path),
                action: {
                  path: path,
                },
              });
            }

            if (entry.isDirectory) {
              items = items.concat(await tree(path));
            }
          }

          return items;
        };

        const newFilename = (await fn.strftime(
          args.denops,
          "%Y/%m/%Y-%m-%d-%H%M%S.",
        )) + args.input;

        let items: Item<ActionData>[] = [{
          word: basename(newFilename),
          display: `[new] ${basename(newFilename)}`,
          action: {
            path: join(dir, newFilename),
          },
        }];
        items = items.concat((await tree(dir)).sort(
          (a, b) => a.word < b.word ? 1 : a.word == b.word ? 0 : -1,
        ));

        controller.enqueue(items);

        controller.close();
      },
    });
  }

  override params(): Params {
    return {};
  }
}
